dvc/repo/__init__.py

```diff
@@ -300,30 +300,32 @@ def graph(self, stages=None, from_directory=None):
         G_active = nx.DiGraph()
         stages = stages or self.stages(from_directory, check_dag=False)
         stages = [stage for stage in stages if stage]
-        outs = []
+        outs = {}
 
         for stage in stages:
             for out in stage.outs:
-                existing = []
-                for o in outs:
-                    if o.path_info == out.path_info:
-                        existing.append(o.stage)
-
-                    in_o_dir = out.path_info.isin(o.path_info)
-                    in_out_dir = o.path_info.isin(out.path_info)
-                    if in_o_dir or in_out_dir:
-                        raise OverlappingOutputPathsError(o, out)
-
-                if existing:
-                    stages = [stage.relpath, existing[0].relpath]
+                if out.path_info in outs:
+                    stages = [
+                        stage.relpath,
+                        outs.get(out.path_info).stage.relpath,
+                    ]
                     raise OutputDuplicationError(str(out), stages)
+                outs[out.path_info] = out
 
-                outs.append(out)
+        for stage in stages:
+            for out in stage.outs:
+                for p in out.path_info.parents:
+                    if p in outs:
+                        raise OverlappingOutputPathsError(
+                            outs.get(p), outs.get(out.path_info)
```

Comment:
`outs.get(out.path_info)` looks like an error, it should be just `out.path_info`. Using `dict.get()` instead of item access concealed it.

Reply:
Also, using
```python
if any(p in outs for p in out.path_info.parents):
    # ...
```
will decrease nesting. A matter of taste for sure.

Reply:
yep,  don't like this nesting ... haven't had time to optimize the style ... will take a second look for sure

I'll fix gets 👍 

`outs.get(out.path_info)` - I think `OverlappingOutputPathsError` expect something internally to get a stage

Reply:
It expects outs. `outs.get(out.path_info)` is the same as `out`, not a stage anyway.

Reply:
check the code inside `OverlappingOutputPathsError`, it runs something like `out.stage` to get it's path

Reply:
Yes, but `outs.get(out.path_info)` is always the same as just `out`.

Reply:
haha ... yep :) will fix it ... btw, `any(something)` does not work - need `p` to get `outs[p]`